Software/src/MoodLampManager.cpp

```diff
@@ -136,39 +141,89 @@ void MoodLampManager::initFromSettings()
 	setCurrentLamp(Settings::getMoodLampLamp());
 }
 
-void MoodLampManager::setCurrentLamp(const int id)
+void MoodLampManager::setCurrentLamp(const QString& moduleName)
 {
 	m_timer.stop();
 
-	if (m_lamp) {
-		delete m_lamp;
-		m_lamp = nullptr;
+	if (!m_jsLamp.isUndefined()) {
+		m_jsLamp = QJSValue();
+		m_jsEngine->collectGarbage();
+	}
+
+	if (!m_lamps.contains(moduleName)) {
+		qWarning() << Q_FUNC_INFO << moduleName << "uknown lamp";
+		return;
 	}
 
-	m_lamp = MoodLampBase::createWithID(id);
+	const MoodLampLampInfo& lampInfo = m_lamps[moduleName];
+
+	const QJSValue& newLamp = m_jsEngine->importModule(lampInfo.modulePath);
+	if (newLamp.isError()) {
+		qWarning() << Q_FUNC_INFO << QString("JS Error in %1:%2 %3")
+			.arg(newLamp.property("fileName").toString())
+			.arg(newLamp.property("lineNumber").toInt())
+			.arg(newLamp.toString());
+		return;
+	}
+	m_lampModuleName = moduleName;
+	m_jsLamp = newLamp;
 	emit moodlampFrametime(1000); // reset FPS to 1
-	if (m_isMoodLampEnabled && m_lamp)
-		m_timer.start(m_lamp->interval());
+	if (m_isMoodLampEnabled && !m_jsLamp.isUndefined())
+		m_timer.start(std::max(1, m_jsLamp.property("interval").toInt()));
```

Comment:
If values larger than X cause errors, it would be nice to either log an error if it is, or add some logic to handle the case that updates the color without running the JS. Not everyone reads comments of other scripts ;)

Reply:
It was to avoid 0 or negative times. I opted for checking on load. Not sure if there's a need for an explicit arbitrary upper limit though (and how much?.. the value will always be within `start()`s range as is)